Sources/MongoSwift/ReadPreference.swift

```diff
@@ -193,104 +186,73 @@ public struct ReadPreference {
-                        " \(MONGOC_SMALLEST_MAX_STALENESS_SECONDS), \(maxStalenessSeconds) given"
+                        " \(MONGOC_SMALLEST_MAX_STALENESS_SECONDS), \(maxStaleness) given"
                 )
             }
-            mongoc_read_prefs_set_max_staleness_seconds(self.readPref, Int64(maxStalenessSeconds))
         }
+        self.mode = mode
+        self.tagSets = tagSets
+        self.maxStalenessSeconds = maxStalenessSeconds
     }
 
-    fileprivate var mode: ReadPreference.Mode {
-        ReadPreference.Mode(mongocMode: mongoc_read_prefs_get_mode(self.readPref))
-    }
+    /// Initializes a new `ReadPreference` by copying a `mongoc_read_prefs_t`. Does not free the original.
+    internal init(copying pointer: OpaquePointer) {
+        self.mode = Mode(mongocMode: mongoc_read_prefs_get_mode(pointer))
 
-    fileprivate var tagSets: [Document]? {
-        guard let bson = mongoc_read_prefs_get_tags(self.readPref) else {
+        guard let tagsPointer = mongoc_read_prefs_get_tags(pointer) else {
             fatalError("Failed to retrieve read preference tags")
         }
         // we have to copy because libmongoc owns the pointer.
-        let wrapped = Document(copying: bson)
+        let wrappedTags = Document(copying: tagsPointer)
+        if !wrappedTags.isEmpty {
+            // swiftlint:disable:next force_unwrapping
+            self.tagSets = wrappedTags.values.map { $0.documentValue! } // libmongoc will always return array of docs
+        }
 
-        guard !wrapped.isEmpty else {
-            return nil
+        let maxStalenessValue = mongoc_read_prefs_get_max_staleness_seconds(pointer)
+        if maxStalenessValue != MONGOC_NO_MAX_STALENESS {
+            self.maxStalenessSeconds = Int(exactly: maxStalenessValue)
         }
+    }
 
+    internal func withMongocReadPreference<T>(body: (OpaquePointer) throws -> T) rethrows -> T {
         // swiftlint:disable:next force_unwrapping
-        return wrapped.values.map { $0.documentValue! } // libmongoc will always give us an array of documents
-    }
+        let rp = mongoc_read_prefs_new(self.mode.mongocMode)! // never returns nil
+        defer { mongoc_read_prefs_destroy(rp) }
+
+        if let tagSets = self.tagSets, !tagSets.isEmpty {
+            let tags = Document(tagSets.map { .document($0) })
+            tags.withBSONPointer { tagsPtr in
+                mongoc_read_prefs_set_tags(rp, tagsPtr)
+            }
+        }
 
-    fileprivate var maxStalenessSeconds: Int? {
-        let maxStaleness = mongoc_read_prefs_get_max_staleness_seconds(self.readPref)
-        return maxStaleness == MONGOC_NO_MAX_STALENESS ? nil : Int(exactly: maxStaleness)
+        if let maxStaleness = self.maxStalenessSeconds {
+            mongoc_read_prefs_set_max_staleness_seconds(rp, Int64(maxStaleness))
+        }
+
+        return try body(rp)
     }
+}
 
-    deinit {
-        mongoc_read_prefs_destroy(readPref)
+internal func withOptionalMongocReadPreference<T>(
```

Comment:
For organizational purposes, would this be better as a static method on `ReadPreference`?

Reply:
that's fine with me, done. we should probably make the same change for the optional document helper for consistency then. I'll do it in a separate PR after this is merged